blog/_posts/2019-01-18-fluxdiffeq.md

```diff
@@ -0,0 +1,822 @@
+  du[2] = dy = (δ*x - γ)*y
+end
+h(p,t) = ones(eltype(p),2)
+prob = DDEProblem(delay_lotka_volterra,[1.0,1.0],h,(0.0,10.0),constant_lags=[0.1])
+
+p = param([2.2, 1.0, 2.0, 0.4])
+params = Flux.Params([p])
+function predict_rd_dde()
+  diffeq_rd(p,reduction,prob,101,MethodOfSteps(Tsit5()),saveat=0.1)
+end
+loss_rd_dde() = sum(abs2,x-1 for x in predict_rd_dde())
+loss_rd_dde()
+```
+
+Additionally we can add randomness to our differential equation to simulate
+how random events can cause extra births or more deaths than expected. This
+kind of equation is known as a stochastic differential equation (SDE).
+Since [DifferentialEquations.jl handles SDEs](http://docs.juliadiffeq.org/latest/tutorials/sde_example.html)
+(and is currently the only library with adaptive stiff and non-stiff SDE integrators),
+these can be handled as a layer in Flux.jl similarly. Here's a neural net layer
+with an SDE:
+
+```julia
+function lotka_volterra_noise(du,u,p,t)
+  du[1] = 0.1u[1]
+  du[2] = 0.1u[2]
+end
+prob = SDEProblem(lotka_volterra,lotka_volterra_noise,[1.0,1.0],(0.0,10.0))
+
+p = param([2.2, 1.0, 2.0, 0.4])
+params = Flux.Params([p])
+function predict_fd_sde()
+  diffeq_fd(p,reduction,101,prob,SOSRI(),saveat=0.1)
+end
+loss_fd_sde() = sum(abs2,x-1 for x in predict_fd_sde())
+loss_fd_sde()
+```
+
+And we can train the neural net to watch it in action and find parameters to make
+the amount of bunnies close to constant:
+
+```julia
+data = Iterators.repeated((), 100)
+opt = ADAM(0.1)
+cb = function ()
+  display(loss_fd_sde())
+  display(plot(solve(remake(prob,p=Flux.data(p)),SOSRI(),saveat=0.1),ylim=(0,6)))
+end
+
+# Display the ODE with the current parameter values.
+cb()
+
+Flux.train!(loss_fd_sde, params, data, opt, cb = cb)
+```
+
+![SDE NN Animation](https://user-images.githubusercontent.com/1814174/51399524-2c6abf80-1b14-11e9-96ae-0192f7debd03.gif)
+
+And we can keep going. There are differential equations
+[which are piecewise constant](http://docs.juliadiffeq.org/latest/tutorials/discrete_stochastic_example.html)
+used in biological simulations, or
+[jump diffusion equations from financial models](http://docs.juliadiffeq.org/latest/tutorials/jump_diffusion.html),
+and the solvers map right over to the Flux.jl neural network frame work through FluxDiffEq.jl
+It is worth noting that FluxDiffEq.jl uses only around ~50 lines of code to pull this all off.
+
+## Implementing the Neural ODE layer in Julia
+
+Let's go all the way back for a second and now implement the neural ODE layer
+in Julia. Remember that this is simply an ODE where the derivative function
+is defined by a neural network itself. To do this, let's first define the
+neural net for the derivative. In Flux, we can define a multilayer perceptron
+with 1 hidden layer and a `tanh` activation function like:
+
+```julia
+dudt = Chain(Dense(2,50,tanh),Dense(50,2))
+```
+
+To define a `neural_ode` layer, we then just need to give it a timespan and use
+the `neural_ode` helper function:

```

Comment:
yes, whatever working makes sense to you here. I usually just use "helper function" as "package function you could've easily written yourself in a few lines"

Reply:
Yeah, but remember that people coming from PyTorch/TF/NumPy assume that if a function is provided in a package it’s because it would be too difficult to implement correctly themselves.